pallets/dactr/src/lib.rs

```diff
@@ -520,6 +641,193 @@ pub mod pallet {
 
 			Ok(().into())
 		}
+
+		#[pallet::call_index(9)]
+		#[pallet::weight(T::WeightInfo::register_blob_offence())]
+		pub fn register_blob_offence(
+			origin: OriginFor<T>,
+			offence_key: OffenceKey,
+			voucher: ValidatorVoucher,
+		) -> DispatchResultWithPostInfo {
+			ensure_none(origin)?;
+
+			// Ensure the key is valid based on the kind and params
+			ensure!(offence_key.is_valid(), Error::<T>::InvalidOffenceKey);
+
+			// Get validators list
+			let validators = T::SessionDataProvider::validators();
+
+			// Check given session
+			let provided_session = voucher.session_index;
+			let current_session = T::ValidatorSet::session_index();
+			ensure!(
+				provided_session == current_session,
+				Error::<T>::InvalidVoucherSession
+			);
+
+			// Checks only run outside of tests and runtime benchmarks
+			let Some(validator_id) =
+				T::AccountId::decode(&mut &voucher.validator.encode()[..]).ok()
+			else {
+				return Err(Error::<T>::InvalidVoucherValidator.into());
+			};
+
+			#[cfg(not(any(test, feature = "runtime-benchmarks")))]
+			{
+				let key = voucher.key;
+				let key_type = sp_core::crypto::key_types::BABE;
+				ensure!(
+					Some(validator_id.clone())
+						== T::SessionDataProvider::get_validator_from_key(key_type, key.encode()),
+					Error::<T>::InvalidVoucherValidator
+				);
+
+				// Check if the extrinsic is coming from an active validator
+				ensure!(
+					validators.contains(&validator_id),
+					Error::<T>::NotAnActiveValidator
+				);
+
+				// Check that the offence key is correctly signed
+				ensure!(
+					voucher.verify_signature((offence_key.clone(), current_session).encode()),
+					Error::<T>::InvalidVoucherSignature
+				);
+			}
+
+			// Get or create the offence
+			let mut record = BlobOffenceRecords::<T>::get(&offence_key).unwrap_or_else(|| {
+				OffenceRecord::<T>::new(
+					offence_key.kind.clone(),
+					offence_key.block_hash,
+					offence_key.blob_hash.clone(),
+					offence_key.missing_validator.clone(),
+				)
+			});
+
+			// If the vouchers already contains my voucher, return an error
+			ensure!(
+				!record
+					.vouches
+					.iter()
+					.any(|v| voucher.validator == v.validator),
+				Error::<T>::DuplicateVouch
+			);
+
+			// If the offence record has not reached threshold, add my boucher
+			let mut added = false;
+			let threshold_reached = if cfg!(test) {
+				false
```

Comment:
Why is this always false for tests? Doesn't it make it impossible to test if threshold_reached even works properly?